### ISSUE TITLE
Improve security of signature generation

### ISSUE THREAD

**Issue by paulmillr** [ethereumjs/ethereumjs-monorepo#3801]
Recently, elliptic.js had a vulnerability, from which it was possible to extract private keys from signatures.

Hedged signatures add additional protections against these private key leaks. noble-curves already support them. More info:

1. https://paulmillr.com/posts/deterministic-signatures/
2. https://ethresear.ch/t/hedged-signatures-ftw/21757

[micro-eth-signer](https://github.com/paulmillr/micro-eth-signer) supports hedged signatures. [Viem](https://viem.sh) also supports them, but it's disabled by default and needs to be enabled using [`setSignEntropy`](https://github.com/wevm/viem/blob/d0ead40b277d0d8094e654d20df903e4e92a551a/src/accounts/utils/sign.ts#L33).

Ethers has [an open issue](https://github.com/ethers-io/ethers.js/issues/4885) for it, they've indicated interest if Geth/Clef added hedging. Geth/Clef have [an open issue](https://github.com/ethereum/go-ethereum/issues/30773).

### Disadvantages

- Signatures (r, s) would become non-deterministic and "new" after every signature.
    - They would still be verifiable. This is not a problem for tests because we can specify our own `random` extraData in tests.

There is no risk for security. If passed-through random is bad, the signature security would be just like today, not worse

**Comment by paulmillr** [comment]
Viem is on board and implemented the changes today.

**Comment by jochem-brouwer** [comment]
Wow, this confuses me a lot. We had a rule in the past (Homestead fork) where we did not allow flipping certain values (namely we do not allow tx signatures where `s > SECP256K1_ORDER_DIV_2`). But now I realize that this is only possible when you know the v/r/s values and you can only end up with one other valid signature which yields the same public key. However when we would allow this extra entropy then each `sign` on the same data with the same private key will yield different `v/r/s` values.

Test script (in `./packages/tx/test/testSig.spec.ts`):

```typescript
import { bytesToHex } from "ethereum-cryptography/utils";
import { createLegacyTx } from "../src/index.js";

const d = createLegacyTx({})
const p = new Uint8Array(32).fill(0x20)

const s = d.sign(p)
console.log(bytesToHex(s.hash()), s.getSenderAddress().toString())
const s2 = d.sign(p)
console.log(bytesToHex(s2.hash()), s2.getSenderAddress().toString())
```

Without the extra entropy, change nothing (as of current master @ e86eace8c6bc4a848c260e6b652fc56e73ac16de). The two signatures will yield the same hash and the same sender, also when started up a different time.

Output:

```
9a94f6841407ebad6d23d4d80396592902847c6bd14fbdcc4f9d99339312d481 0xb6e610921b0a0f6f608c0e1f29a845552bc6db2c
9a94f6841407ebad6d23d4d80396592902847c6bd14fbdcc4f9d99339312d481 0xb6e610921b0a0f6f608c0e1f29a845552bc6db2c
```

To add entropy, change `./packages/util/src/signature.ts`:

```typescript
export function ecsign(
  msgHash: Uint8Array,
  privateKey: Uint8Array,
  chainId?: bigint,
): ECDSASignature {
  const sig = secp256k1.sign(msgHash, privateKey, { extraEntropy: true })
  const buf = sig.toCompactRawBytes()
  ```
  
  Each time sign is run (also thus the first and second signature on the run) the hash will change but the sender will not.
  
  Sample output:
  
  ```
  eacc68ac9306c1c7d53ef351257b470dbfbede13cf29d315f8c1ac2e7c222d72 0xb6e610921b0a0f6f608c0e1f29a845552bc6db2c
d02712f4cdd474503f4dda38b6bafe9f1721f551a9cc51523acee868b15f25f5 0xb6e610921b0a0f6f608c0e1f29a845552bc6db2c
```

Forgive my novice-ness in cryptography here, but if we add the entropy, then this means that each time if someone uses our libraries and uses the same private key and the same data to sign (i.e. the same transaction), then it will yield a different `v/r/s` scheme and thus also a different hash. Is this wanted? :thinking: (Or am I doing something wrong here?)

**Comment by paulmillr** [comment]
It is not default because it would break tests, which assume deterministic outputs.

It is also not clear which user applications rely on deterministic outputs. But I think it's safe to change in the upcoming major release while keeping a changelog **NOTE**.

Using extraEntropy would massively harden security of cryptography. Again, to prove my point, soon a vulnerability in popular library (not noble) would be disclosed. I can send you more info on it in direct messages. It looks bad and I want to prevent such things in advance.

**Comment by paulmillr** [comment]
Besides `ecsign`, EIP712 / EIP191 messages would also need to be edited to use entropy.

By "not clear which user applications rely on deterministic outputs" I mean the following thing:

- End-user applications which verify tx hash or EIP712 sig **do not** know which tx hash is produced
    - The applications can **only** verify the fact the signature is valid
    - Nothing is changed for them
- The only place where different signatures may affect something is deep applications, maybe EVM? with access to private keys and such

**Comment by jochem-brouwer** [comment]
Hi Paul, we will address this in the upcoming major release (which we plan to do ASAP, also because Pectra is coming up!).

Is this issue related: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3872 and does that also link the vuln mentioned you mentioned in this comment: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3801#issuecomment-2487934982? https://github.com/advisories/GHSA-vjh7-7g9h-fjfh

**Comment by paulmillr** [comment]
yes. sorry. I forgot

**Comment by jochem-brouwer** [comment]
Hi Paul, I have studied the articles you linked and one question: is adding the `extraEntropy: true` option to the relevant methods the same as the "hedged signatures" proposal, or is "only" switching out "deterministic" signatures for "random" signatures?  (I'm not entirely clear in this context if those are the same, or different?)

**Comment by paulmillr** [comment]
extraEntropy: true Is hedged. Noble does not support fully random

**Comment by acolytec3** [comment]
Just to make sure we're clear on what's happening, `micro-eth-signer` is already doing `extraEntropy: true` by default?  

What about `ethereum-cryptography@3.1.0`?  We're currently importing `secp256k1` from there so do we need to change anything other than expose the parameter in our `sign` function?

**Comment by paulmillr** [comment]
No, it is not default right now. It has been optional, but the option was there for many years. I can make it default if you also make it. Will also ask Viem.

I don't think eth-cryptography needs any changes.

**Comment by acolytec3** [comment]
Ok, cool, we'll just ensure we pass in the `extraEntropy` parameter going forward.

**Comment by paulmillr** [comment]
https://github.com/wevm/viem/issues/3355

**Comment by jochem-brouwer** [comment]
I'm setting `extraEntropy: true` by default with override options, see: https://github.com/ethereumjs/ethereumjs-monorepo/pull/3873

**Comment by paulmillr** [comment]
@szaglam This is NOT breaking RFC 6979. The change is a part of RFC 6979.

Changes in ethereumjs are opt-out which means you can still use old methods.

**Comment by paulmillr** [comment]
Your point is invalid:

1. It was made deterministic because pure randomness can't be relied upon after PS3 hack.
2. It is still deterministic even with 3.6 randomization if you specify `randomness=00000`.
3. The change does not make signatures rely on randomness generator (which was the reason 6979 was created). If randomness generator is bad, it would fall back "worst case" to determinism.
4. The 3.6 randomization has been present in [libsecp256k1](https://github.com/bitcoin-core/secp256k1) since 2015.

**Comment by holgerd77** [comment]
We have discussed this with Go Ethereum and will switch back to keep the deterministic behavior as default, and have the hedged version as opt in. Otherwise this will get too breaking and is too disruptive regarding interoperability between clients, if e.g. tx hashes are not directly comparable anymore.

**Comment by paulmillr** [comment]
@holgerd77 that's fine, I will work more on convincing community for the future releases.

> tx hashes are not directly comparable anymore

What do you mean? This seems like a thing for tests. In production, one can't really compare hashes, since privkeys aren't known, so tx hashes can't be verified for equality.

**Comment by szaglam** [comment]
Thank you! If ethereumjs changed this behavior, then many wallet derived keys were going to be invalid on MetaMask for the protocols above that I know of, and potentially many more. Really appreciate it!
